vault/diagnose/diagnose_tls_test.go

```diff
@@ -0,0 +1,146 @@
+					TLSCertFile:                   "./certs/server.crt",
+					TLSKeyFile:                    "./certs/server.key",
+					TLSClientCAFile:               "./certs/rootca.crt",
+					TLSMinVersion:                 "tls11",
+					TLSRequireAndVerifyClientCert: true,
+					TLSDisableClientCerts:         true,
+				},
+				{
+					Type:                          "tcp",
+					Address:                       "127.0.0.1:443",
+					ClusterAddress:                "127.0.0.1:8201",
+					TLSCertFile:                   "./certs/server2.crt",
+					TLSKeyFile:                    "./certs/server2.key",
+					TLSClientCAFile:               "./certs/rootca2.crt",
+					TLSMinVersion:                 "tls12",
+					TLSRequireAndVerifyClientCert: true,
+					TLSDisableClientCerts:         false,
+				},
+				{
+					Type:                          "tcp",
+					Address:                       "127.0.0.1:443",
+					ClusterAddress:                "127.0.0.1:8201",
+					TLSCertFile:                   "./certs/server3.crt",
+					TLSKeyFile:                    "./certs/server3.key",
+					TLSClientCAFile:               "./certs/rootca3.crt",
+					TLSMinVersion:                 "tls13",
+					TLSRequireAndVerifyClientCert: false,
+					TLSDisableClientCerts:         true,
+				},
+			},
+		},
+	}
+	conf := &vault.CoreConfig{
+		RawConfig: serverConf,
+	}
+	core := vault.TestCoreWithConfig(t, conf)
+	return core
+}
+
+func TestTLSValidCert(t *testing.T) {
+	listeners := []listenerutil.Listener{
+		{
+			Config: &configutil.Listener{
+				Type:                          "tcp",
+				Address:                       "127.0.0.1:443",
+				ClusterAddress:                "127.0.0.1:8201",
+				TLSCertFile:                   "./../../api/test-fixtures/keys/cert.pem",
+				TLSKeyFile:                    "./../../api/test-fixtures/keys/key.pem",
+				TLSClientCAFile:               "./../../api/test-fixtures/root/rootcacert.pem",
+				TLSMinVersion:                 "0",
+				TLSRequireAndVerifyClientCert: true,
+				TLSDisableClientCerts:         false,
+			},
+		},
+	}
+	err := ListenerChecks(listeners)
+	if err != nil {
+		t.Fail()
+	}
+}
+
+func TestTLSFakeCert(t *testing.T) {
+	listeners := []listenerutil.Listener{
+		{
+			Config: &configutil.Listener{
+				Type:                          "tcp",
+				Address:                       "127.0.0.1:443",
+				ClusterAddress:                "127.0.0.1:8201",
+				TLSCertFile:                   "./test-fixtures/fakecert.pem",
+				TLSKeyFile:                    "./../../api/test-fixtures/keys/key.pem",
+				TLSClientCAFile:               "./../../api/test-fixtures/root/rootcacert.pem",
+				TLSMinVersion:                 "0",
+				TLSRequireAndVerifyClientCert: true,
+				TLSDisableClientCerts:         false,
+			},
+		},
+	}
+	err := ListenerChecks(listeners)
+	if err == nil {
+		t.Errorf("TLS Config check on fake certificate should fail")
+	}
+	if err.Error() != "tls: failed to find any PEM data in certificate input" {
```

Comment:
is this error "type" included in some library we're using under the hood, or does it need to be a string?

Reply:
It's already of type string.

Reply:
i'm wondering if you could do some error comparison like

`if err == storage.ErrObjectNotExist {
	return nil, nil
}`

rather than directly comparing to a string. like does the library have something like this defined

`var (
	// ErrBucketNotExist indicates that the bucket does not exist.
	ErrBucketNotExist = errors.New("storage: bucket doesn't exist")
	// ErrObjectNotExist indicates that the object does not exist.
	ErrObjectNotExist = errors.New("storage: object doesn't exist")
)`

Reply:
oops that formatting is awful, sorry!

Reply:
Yeah, I know what you mean, but in this case I can't do it. The error is `return fail(errors.New("tls: failed to find any PEM data in certificate input"))` in src/crypto/tls/tls.go.